content/blog/managing-aws-credentials-on-cicd/index.md

```diff
@@ -0,0 +1,231 @@
+
+However, if you want to automate changes to your AWS infrastructure from a 3rd party CI/CD service, you must provide those credentials. Here is where following security best practices come into play.
+
+If we do need to provide credentials to a 3rd party, rather than hoping it will be 100% secure forever (which is impossible), we can instead make those credentials volatile. If we regularly invalidate and rotate the credentials given supplied to CI/CD system, we can dramatically reduce the impact of any accidental disclosure. Since even if the IAM User’s credentials were leaked, by the time they were discovered and used, they would no longer be valid. We’ll go into how to manage key rotation shortly.
+
+### Properly Storing Credentials
+
+While key rotation may thwart a compromised credential, we still need to take precautions to ensure that they are correctly stored.
+
+Your CI/CD provider likely has built-in support for handling “secrets” or “secure variables.” Typically these are user-supplied configuration values that are write-only and only accessible by the CI/CD worker jobs at runtime. For example, Travis CI has [encrypted environment variables](https://circleci.com/docs/2.0/env-vars/#overview) or [secure environment variables](https://circleci.com/docs/2.0/env-vars/#overview) in CircleCI.
+
+## Automate Rotating and Revoking those Credentials
+
+When a CI/CD system has access to an IAM User’s AWS credentials, we need to take steps to rotate those credentials regularly, even if we trust the CI/CD provider to secure them properly.
+
+> This is one of the advantages of running your CI/CD system within your AWS account. If your deployment workers are simply EC2 instances, you wouldn’t need to share IAM User credentials. The EC2 instances themselves [can be configured to assume an IAM Role](https://aws.amazon.com/blogs/security/new-attach-an-aws-iam-role-to-an-existing-amazon-ec2-instance-by-using-the-aws-cli/).
+>
+> The details about how IAM Roles are a more secure alternative to IAM User credentials are covered in the next section.
+
+“Just create a new IAM User credential, log into the CI/CD system, and then update the job’s settings” sounds simple enough. But it’s likely going to be tedious, and if you need to repeat the process for dozens if not hundreds of CI/CD pipelines, you need a better solution. Unless you start with an automated approach for key rotation, you are unlikely to add one later.
+
+The AWS security blog describes how to [rotate access keys for IAM Users](https://aws.amazon.com/blogs/security/how-to-rotate-access-keys-for-iam-users/).
+
+1. Create a second access key in addition to the one in use.
+1. Update all your applications to use the new access key and validate that the applications are working.
+1. Change the state of the previous access key to inactive.
+1. Validate that your applications are still working as expected.
+1. Delete the inactive access key.
+
```

Comment:
It seems like I first need to dust off the code (collecting dust over at https://github.com/pulumi/autobots/pull/7) and publish it somewhere. (Maybe in the examples repo?) And include snippets or a link in the text here.

Ultimately this step needs to be a "turn key" type experience to be useful.

Reply:
I want to split the post into two parts. This part discusses what needs to happen and the second part would be an implementation example.

Reply:
let's save this for a followup post